### PR TITLE
MRG: Fix GDF

### LINKED ISSUE #5604: Negative meas_date in gdf2 example
I was not expecting 'meas_date' to be negative. Is this a bug? or that this example indeed has negative meas_date?

#### Steps and/or code to reproduce
```py
import os.path as op
from mne.io import read_raw_edf
from mne.annotations import _handle_meas_date
from mne.datasets import testing

data_path = testing.data_path(download=False)
gdf2_path = op.join(data_path, 'GDF', 'test_gdf_2.20')

raw = read_raw_edf(gdf2_path + '.gdf', eog=None, misc=None, preload=True,
                    stim_channel='STATUS')
assert _handle_meas_date(raw.info['meas_date']) > 0
```

### PR DESCRIPTION
**Pull request by larsoner**
Closes #5604.

Also fixes a bug where our slicing logic in `base.py` was not safe for when `raw[:n_ch + 1]` was done. It was exposed by adding `_test_raw_reader` to `test_gdf.py`, where there were only 17 channels but `_test_raw_reader` looked at `raw[:20]`. Now it works like Python slicing where this still gives all 17 channels, but does not throw an error.

### CODE REVIEW

**Comment by larsoner** [comment]
@massich let me know if you want to merge this and rebase #5574 or if you want to merge #5574 and I'll rebase this, I'm happy either way

**Comment by massich** [comment]
I've no problem either. I can review this one, we merge it and I rebase. That's fine.

File: mne/io/base.py [hunk 1]
```diff
@@ -467,6 +467,7 @@ def _read_segment(self, start=0, stop=None, sel=None, data_buffer=None,
 
         #  Initialize the data and calibration vector
         n_sel_channels = self.info['nchan'] if sel is None else len(sel)
+        assert n_sel_channels <= self.info['nchan']
```
Comment (massich) [inline mne/io/base.py]: do you want to keep this assert?

File: doc/whats_new.rst [hunk 2]
```diff
@@ -99,6 +99,8 @@ Bug
 
 - Fix bug with channel names in ``mgh70`` montage in :func:`mne.channels.read_montage` by `Eric Larson`_
 
+- Fix bug in :func:`mne.io.read_raw_edf` where GDF files had ``info['highpass']`` and ``info['lowpass']`` set to NaN by `Eric Larson`_
```
Comment (massich) [inline doc/whats_new.rst]: do you want to add that `meas_date` has also changed?

File: mne/io/edf/edf.py [hunk 3]
```diff
@@ -896,11 +902,10 @@ def _read_gdf_header(fname, stim_channel, exclude):
             meas_id['loc'] = loc
 
             date = np.fromfile(fid, np.uint64, 1)[0]
-            if date == 0:
-                date = datetime.datetime(1, 1, 1)
-            else:
+            if date != 0:
                 date = datetime.datetime(1, 1, 1) + \
```
Comment (massich) [inline mne/io/edf/edf.py]: So you set `meas_date = DATE_NONE` in the top, and if something is read, then you overwrite it.

This snipped appears more than once maybe we can make a function:
```py
def _compute_edf_gdf_meas_date(date):
   if date == 0: 
       return DATE_NONE
   else:
      meas_date = datetime.datetime(1, 1, 1) + \
	    		   datetime.timedelta(date * pow(2, -32) - 367)
      meas_date (calendar.timegm(meas_date.utctimetuple()), 0)
      return meas_date
```

File: mne/io/base.py (hunk 1 above)
Comment (larsoner) [reply mne/io/base.py]: yes it's like an "active comment" (plus prevents us from doing something dumb in a calling function when developing)

File: doc/whats_new.rst (hunk 2 above)
Comment (larsoner) [reply doc/whats_new.rst]: ahh yes I should add it

File: mne/io/edf/edf.py (hunk 3 above)
Comment (larsoner) [reply mne/io/edf/edf.py]: I think a function for this is overkill

**Comment by larsoner** [comment]
OSX failure is unrelated, I'll go ahead and merge so you can rebase your PR

File: mne/io/base.py (hunk 1 above)
Comment (massich) [reply mne/io/base.py]: Don't get me wrong, I'm totally in favor of that kind of asserts.

**Comment by agramfort** [comment]
thx @larsoner !

**Pull request merged**